File: tests/leetcode/x/lua/0044.lua
local function isMatch(s, p)
  local i, j, star, match = 1, 1, -1, 1
  while i <= #s do
    if j <= #p and (p:sub(j, j) == '?' or p:sub(j, j) == s:sub(i, i)) then
      i = i + 1; j = j + 1
    elseif j <= #p and p:sub(j, j) == '*' then
      star = j; match = i; j = j + 1
    elseif star ~= -1 then
      j = star + 1; match = match + 1; i = match
    else
      return false
    end
  end
  while j <= #p and p:sub(j, j) == '*' do j = j + 1 end
  return j > #p
end

local lines = {}
for line in io.lines() do
  local cleaned = line:gsub("\r", "")
  table.insert(lines, cleaned)
end
if #lines == 0 or lines[1]:match('^%s*$') then return end
local idx = 1
local t = tonumber(lines[idx]); idx = idx + 1
local out = {}
for _ = 1, t do
  local n = tonumber(lines[idx]); idx = idx + 1
  local s = n > 0 and lines[idx] or ''
  if n > 0 then idx = idx + 1 end
  local m = tonumber(lines[idx]); idx = idx + 1
  local p = m > 0 and lines[idx] or ''
  if m > 0 then idx = idx + 1 end
  table.insert(out, isMatch(s, p) and 'true' or 'false')
end
io.write(table.concat(out, '\n'))

File: tests/leetcode/x/swift/0044.swift
import Foundation

func isMatch(_ s: String, _ p: String) -> Bool {
  let ss = Array(s)
  let pp = Array(p)
  var i = 0
  var j = 0
  var star = -1
  var match = 0
  while i < ss.count {
    if j < pp.count && (pp[j] == Character("?") || pp[j] == ss[i]) {
      i += 1
      j += 1
    } else if j < pp.count && pp[j] == Character("*") {
      star = j
      match = i
      j += 1
    } else if star != -1 {
      j = star + 1
      match += 1
      i = match
    } else {
      return false
    }
  }
  while j < pp.count && pp[j] == Character("*") { j += 1 }
  return j == pp.count
}

let lines = String(data: FileHandle.standardInput.readDataToEndOfFile(), encoding: .utf8)?
  .split(separator: "\n", omittingEmptySubsequences: false)
  .map { String($0).replacingOccurrences(of: "\r", with: "") } ?? []

if !lines.isEmpty && !lines[0].trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
  var idx = 0
  let t = Int(lines[idx].trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
  idx += 1
  var out: [String] = []
  for _ in 0..<t {
    let n = Int(lines[idx].trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
    idx += 1
    let s = n > 0 ? lines[idx] : ""
    if n > 0 { idx += 1 }
    let m = Int(lines[idx].trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
    idx += 1
    let p = m > 0 ? lines[idx] : ""
    if m > 0 { idx += 1 }
    out.append(isMatch(s, p) ? "true" : "false")
  }
  print(out.joined(separator: "\n"), terminator: "")
}
